### DISCUSSION TITLE
Arbitrary commands

### DISCUSSION THREAD

**Discussion by atwright147** [photostructure/exiftool-vendored.js#90] [Q&A] [answered]
Hey,

Is it possible to run arbitrary command through exiftool-vendored, such as something like (obvs without the call to `exiftool` and I know I can supply my own config: `exiftool -config my.config -b -rotatedthumbnail image.ARW > out.jpg` (https://exiftool.org/forum/index.php?topic=9380.msg63389#msg63389)

I can't see a way to call `extractBinaryTag()` with any args

**Comment by mceachen** [comment, marked as answer]
Your `-config` option can be provided at launch with the `exiftoolArgs` constructor option.

Note that the instance of `ExifTool` will use that `-config` for all calls made to that instance, though!

**Reply by atwright147** [reply]
Awesome and how would I use `-b -rotatedthumbnail`?

**Reply by mceachen** [reply]
Ah crap I totally missed the `-rotatedthumbnail` arg, sorry!

The `tagname` argument in the `BinaryExtractionTask.for` method is what you're looking for, which is proxied through the `extractBinaryTag` method. This _should_ work (untested, though):

```ts
import { DefaultExiftoolArgs, ExifTool } from 'exiftool-vendored';

const et = new ExifTool({exiftoolArgs: ["-config", "my.config", ...DefaultExiftoolArgs]})
await et.extractBinaryTag("rotatedthumbnail", "image.ARW", "out.jpg")
```

For other, more exotic command lines, though, you'll want to create your own subclass of `ExifToolTask`. Look at `ReadTask` and `BinaryExtractionTask` for examples.

Edit 2021-02-04: added `DefaultExiftoolArgs` (derp!)

**Reply by atwright147** [reply]
That's amazing! Thank you I will have a go ASAP :)

**Reply by atwright147** [reply]
I have been trying to get it to work but I get nothing (not even an error message). It is probably me being stupid.

I have put together a self contained demo (including a RAW file from raw.pixls.us. Would you mind having a go with it and seeing what I might be doing wrong?

https://github.com/atwright147/exiftool-vendored-preview-orientation-test

**Reply by mceachen** [reply]
~~Your code looks fine, but you're trying to extract a binary tag that doesn't exist on that file:~~

```sh
$ exiftool -j DSC01005.ARW | grep -i binary
  "TiffMeteringImage": "(Binary data 2640 bytes, use -b option to extract)",
  "PreviewImage": "(Binary data 902424 bytes, use -b option to extract)",
  "ThumbnailImage": "(Binary data 9693 bytes, use -b option to extract)",
```

~~(there's very little standardization to tags, which is why I have to [fetch from so many different tags just for the "captured-at" time](https://photostructure.com/faq/captured-at/)).~~


Edit: ah, the RotatedThumbnail is from the config. Gotcha. Poking at it now.

This works (unsurprisingly):

```sh
exiftool -config my.config DSC01005.ARW -RotatedThumbnail -b > out.jpg
```

This also works as expected (which is the command exiftool-vendored should be running):

```sh
printf -- '-b\n-RotatedThumbnail\n./DSC01005.ARW\n-w\n%%0fout.jpeg\n-ignoreMinorErrors\n-execute\n' | \
  exiftool -config my.config -stay_open True -@ -
```

**Reply by atwright147** [reply]
So, what is going wrong?

**Reply by mceachen** [reply]
Something seems to go south right after the `enqueueTask` call: it's like node just decides to shut down at that point. If you add a console.log after your try/catch it doesn't get there.

Using the global `exiftool` (that doesn't have a configPath) doesn't behave like this, fwiw.

**Reply by mceachen** [reply]
Ugh: `exiftool-vendored` assumes `exiftool` is spun up with the `-stay_open` option.

When you say `exiftoolArgs: ["-config", configPath]`, _that becomes the only arguments given to ExifTool_.

The solution is to change line 13:

```
$ git diff
diff --git a/index.ts b/index.ts
index 0bd0796..6c57f07 100755
--- a/index.ts
+++ b/index.ts
@@ -1,7 +1,7 @@
 #!npx ts-node
 
 import path from 'path';
-import { ExifTool } from 'exiftool-vendored';
+import { DefaultExiftoolArgs, ExifTool } from 'exiftool-vendored';
 
 const configPath = path.resolve('./my.config');
 const sourceImagePath = path.resolve('./DSC01005.ARW');
@@ -10,7 +10,7 @@ const destImagePath = path.resolve('./out.jpg');
 console.info({ configPath, sourceImagePath, destImagePath });
 
 (async () => {
-  const et = new ExifTool({exiftoolArgs: ['-config', configPath]});
+  const et = new ExifTool({exiftoolArgs: ['-config', configPath, ...DefaultExiftoolArgs]});
   try {
     await et.extractBinaryTag('rotatedthumbnail', sourceImagePath, destImagePath);
   } catch (err) {
```

I just edited the first code snippet to be correct as well.

**Reply by atwright147** [reply]
Nice debugging!

Is this something you will be adding to the package?

**Reply by atwright147** [reply]
Then I just need to figure out how to amend that config to work with JpegFromRaw and Previews to get what I am actually after :)

This will be super useful in conjunction with CSS `image-orientation`

**Reply by mceachen** [reply]
> Is this something you will be adding to the package?

No. I could throw an error or emit a warning if `exiftoolArgs` doesn't include  `-stay_open True`, though.

> This will be super useful in conjunction with CSS image-orientation

https://bugs.chromium.org/p/chromium/issues/detail?id=158753 may be interesting (and there are issues on mobile browsers as well).

I had to make [PhotoStructure](https://photostructure.com/faq/why-photostructure/) do all image rotations server-side :cry: because of these issues.

**Reply by atwright147** [reply]
But if I do this, the following, it will work?:

```ts
const et = new ExifTool({exiftoolArgs: ["-config", "my.config", ...DefaultExiftoolArgs]})
await et.extractBinaryTag("rotatedthumbnail", "image.ARW", "out.jpg")
```

**Reply by atwright147** [reply]
I have no idea why I asked that lol. I just tried it and it worked.

That suits me. Thank you :)
